opsi/modules/color.py

```diff
@@ -208,3 +211,117 @@ class Outputs:
+        draw_hsv: bool
+
+    @dataclass
+    class Inputs:
+        img: Mat
+
+    @dataclass
+    class Outputs:
+        color: Color
+        img: Mat
+
+    def run(self, inputs):
+        # Find the pixel coordinates to sample in the image
+        height, width = inputs.img.shape[:2]
+        sample_coords = (
+            int(width * self.settings.x_pct / 100.0 + 10),
+            int(height * self.settings.y_pct / 100.0 + 10),
+        )
+        color_bgr = inputs.img[sample_coords[1], sample_coords[0]]
+        draw = None
+        if self.settings.draw_color:
+            draw = np.copy(inputs.img.mat)
+            # Draw a small circle (of radius 5) to show the point.
+            cv2.circle(draw, sample_coords, 5, (0, 0, 255), 3)
+
+            # Find the color in HSV to make a contrasting color
+            color_hsv = cvw.bgr_to_hsv(np.uint8([[color_bgr]]).view(Mat))[0][0]
+
+            color_hsv[0] *= 2  # Scale the hue value to be in a range of 0-359
+
+            # Create a string to represent the color in either RGB or HSV
+            if self.settings.draw_hsv:
+                color_str = "H{} S{} V{}".format(*color_hsv)
+            else:
+                color_str = "B{} G{} R{}".format(*color_bgr)
+
+            # Choose a (Hopefully) Contrasting color
+            draw_color = (
+                int(255 - color_bgr[0]),
+                int(255 - color_bgr[1]),
+                int(255 - color_bgr[2]),
+            )
+
+            cv2.putText(
+                draw,
+                color_str,
+                (sample_coords[0] + 10, sample_coords[1] + 10),
+                cv2.FONT_HERSHEY_SIMPLEX,
+                0.5,
+                draw_color,
+                lineType=cv2.LINE_AA,
+            )
+
+            draw = draw.view(Mat)
+
+        color = Color(color_bgr[2], color_bgr[1], color_bgr[0])
+        return self.Outputs(color=color, img=draw)
+
+
+class ColorDetector(Function):
+    @dataclass
+    class Settings:
+        red_hue: Slide(min=0, max=359, decimal=False)
+        yellow_hue: Slide(min=0, max=359, decimal=False)
+        green_hue: Slide(min=0, max=359, decimal=False)
+        blue_hue: Slide(min=0, max=359, decimal=False)
+
+    @dataclass
+    class Inputs:
+        color: Color
+
+    @dataclass
+    class Outputs:
+        color_string: str
+
+    def run(self, inputs):
+        def hue_dist(test: int, reference: int):
+            return min(abs(reference - test), abs(reference + 360 - test))
+
+        color_hue = (
+            cvw.bgr_to_hsv(
+                np.uint8(
+                    [[[inputs.color.blue, inputs.color.green, inputs.color.red]]]
+                ).view(Mat)
+            )[0][0][0]
+            * 2
+        )
+
+        hue_strings = {
+            self.settings.red_hue: "R",
+            self.settings.yellow_hue: "Y",
+            self.settings.green_hue: "G",
+            self.settings.blue_hue: "B",
+        }
+
+        output_str = ""
```

Comment:
Rather than creating an empty string here, it would be better to create a default output (such as "No Color") or just returning None rather than an empty string.